docs/RFCS/20210722_on_update_and_row_rehoming.md

```diff
@@ -0,0 +1,250 @@
+region and falling back to the user’s primary region in cases where the gateway
+region is not a region in the database.
+
+This design presents challenges in the case where the user performs an INSERT in
+a region that is not their “home”. For example, if an American user was to
+INSERT their data in Europe then travel back to America, their data would stay
+homed in Europe, defeating the purpose of this optimization.
+
+This design proposes the addition of an ON UPDATE constraint to allow users to
+specify an expression to be applied on UPDATE/UPSERT. This extension will then
+be used to recalculate `crdb_region` from user’s gateway region on UPDATE/UPSERT
+with the idea that continuing to home rows post-insert will lead to more
+efficient row placement.
+
+# Technical design
+
+Note: In this RFC, the term UPDATE is used to refer to behavior that should
+happen on both UPDATE and UPSERT.
+
+## Update Mechanism
+
+To support rehoming, a "recalculate on UPDATE" mechanism will be implemented.
+This mechanism will allow users to mark a column with an expression to be
+recalculated whenever a row is modified. This mechanism can be thought of as
+similar to DEFAULT expressions in that if a column is not explicitly set in an
+UPDATE, it will be filled with the value calculated from the column's ON UPDATE
+expression.
+
+Note that this ON UPDATE expression is distinct from a column's DEFAULT
+expression i.e., a user can set different DEFAULT and ON UPDATE expressions.
+
+### Interaction with CASCADE and backfills
+
+If a row in a table with an ON UPDATE column is modified by an ON UPDATE CASCADE
+operation, the ON UPDATE expression will be applied. This decision is made
+primarily in order to be semantically consistent with ON UPDATE in Postgres
+triggers; Postgres ON UPDATE triggers fire when a row is modified via an ON
+UPDATE CASCADE. This decision will become particularly relevant if triggers are
+implemented in Cockroach in the future.
+
+This behavior can also be motivated via the row rehoming use case. If a REGIONAL
+BY ROW table `rt` has a foreign key onto another RBR table `ft`, we'd like rows
+in `rt` to be in the same region as the rows they reference in `ft`. Therefore,
+an update to a key in `ft` should require that the region for referencing rows
+in `rt` be recalculated, leading to rows in `rt` being in the same region as the
+rows they reference in `ft`. See the rehoming section below for more details on
+how ON UPDATE will be used in REGIONAL BY ROW tables.
+
+In the case of backfills, however, the ON UPDATE clause will not be
+re-evaluated. Applying the ON UPDATE to all rows is likely unexpected behavior,
+particularly when the ON UPDATE captures something unique about each row
+(region, last modified timestamp, etc.).
+
+### Syntax
+
+Since this addition is a constraint on a column, it can be specified at create
+time with the following syntax:
+
+```
+<column_name> <column_type> <other_constraints> ON UPDATE <update_expr>
```

Comment:
PostgreSQL's dialect already has opinions about what comes after `ON`. 

The general syntax is `ON <eventtype> [<constraintname>] DO <action> <args...>`

So you probably want this:

`ON UPDATE [<constrainname>] DO SET ...`

Either `SET <set_clause>` (`set_clause` is alreayd defined, see grammar) or `SET TO  <expr>`

Reply:
Example: `ON UPDATE DO SET x = 123` or `ON UPDATE upd_cstr DO SET x = 123`

Reply:
@knz do you have a source for this general purpose?

i think having a constraint name for it is valid (not sure how this would look on the constraint table though, would need extra meta to convey it's kind of a trigger?), but i would argue supporting `ON UPDATE SET ...` on top of `ON UPDATE [constraint_name] DO SET ...` seems reasonable, without the DO in between if there is no constraint name in between. the extra DO (so far) seems a bit extra, but i can be convinced otherwise.

Reply:
1) the SET is pretty much a given.

See the SQL standard, part 2 (Foundation) for foreign key relations:
![image](https://user-images.githubusercontent.com/642886/127299683-aa9ba551-5efc-4e66-9de4-1880da497610.png)

and the pg doc for INSERT
![image](https://user-images.githubusercontent.com/642886/127300049-b5388778-7ffb-4b14-8f2e-dfcf5eb17fc1.png)

2) the "DO" is quite important if we allow an optional identifier (e.g. a constraint name) after ON UPDATE; without mandating "DO" the grammar would be ambiguous.

Reply:
one thing i'd add here is that it's quite possible PG will add `ON UPDATE` in a future release. (i'm guessing based on other features they have added for parity, like computed columns.)

obviously we can't predict what syntax they might use, but we should try to pick something which is less likely to create an amibiguity with what they do. so this might be a point in favor of `DO UPDATE SET`

Reply:
I think mirroring the foreign key syntax here makes sense particularly because we aren't accepting arbitrary columns to be set; we're only allowing the user to specify an expression to be evaluated and set ON UPDATE.

Reply:
We did consider having a single expression, and the original working doc used one. We ultimately decided to move to separate clauses for 2 reasons:

1. Aligning with MySQL - MySQL has [ON UPDATE](https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html) syntax that this feature borrows from. We thought it best to match the MySQL syntax where it made sense to make it easy to understand this feature.
2. Use Case - There is a use case, particularly when you want to not apply the expression until an UPDATE is received, like a `last_updated_timestamp` column. For example, you could have `DEFAULT NULL ON UPDATE CURRENT_TIMESTAMP` to not mark the last updated timestamp until an UPDATE is actually issued.